### ISSUE TITLE
parameters turned to None after jvp of vjp

### ISSUE THREAD

**Issue by pablo2909** [patrick-kidger/equinox#578]
Hi,

I am playing with VJP and JVP and implemented two equivalent algorithm. The first one `compute_vjpjvp` applies a jvp on a function and then vjp the whole, the second one `compute_jvpvjp` is the inverse. Please notice in the code below that the jvp is applied to one argument of the function and the vjp is applied to an other argument.

```python
import jax
import jax.numpy as jnp

@jax.jit
def f(x, y):
    h = jnp.linalg.norm(x) * jnp.linalg.norm(y)
    for _ in range(20):
        x *= jnp.sin(h*x)
    return x

x0 = jnp.arange(1., 4.)   # (3,)
y0 = jnp.arange(1., 6.)   # (5,)
z0 = jnp.arange(2., 5.)   # (3,)

def compute_vjpjvp(z, x1, y1):

    def Dfz(y):
        f_xy, jvp =  jax.jvp(lambda s: f(s, y), (x1,), (z,))
        return jvp
    
    return jax.vjp(Dfz, y1)[1](f(x1, y1))[0]

vjpjvp = compute_vjpjvp(z0, x0, y0)

def compute_jvpvjp(z, x1, y1):
    def xDf(x):
        f_xy, vjpfun = jax.vjp(lambda s: f(x, s), y1)
        return vjpfun(jax.lax.stop_gradient(f_xy))
    return jax.jvp(xDf, (x1,), (z,))[1][0]

jvpvjp = compute_jvpvjp(z0, x0, y0)

print(vjpjvp == jvpvjp)
```

We can check that both functions return the same result.

However if we apply those function on an `MLP` it turns out that `compute_jvpvjp` turns the bias of the last layer to `None`. 

```python

import equinox as eqx
from jax import random
score = eqx.nn.MLP(in_size=3, out_size=3, width_size=10, depth=2, key=random.PRNGKey(0))

def compute_jvpvjp(z, x1, score):
    def xDf(x):
        f_xy, vjpfun = eqx.filter_vjp(
            lambda s: s(x),
            score,
        )
        return vjpfun(jax.lax.stop_gradient(f_xy))

    return eqx.filter_jvp(xDf, (x1,), (z,))[1][0]

def compute_vjpjvp(z, x1, score):
    def Dfz(score):
        f_xy, jvp = eqx.filter_jvp(
            lambda s: score(s), (x,), (z,)
        )
        return jvp

    return eqx.filter_vjp(Dfz, score)[1](score(x1))[0]

z = jnp.arange(3.)
x = jnp.arange(1., 4.)
print(compute_vjpjvp(z, x, score).layers[2].bias)
print(compute_jvpvjp(z, x, score).layers[2].bias)
```

Why is that ?

Thank you for any help :)

**Comment by lockwo** [comment]
What is `f` in the original problem? When I copy the jax code, it is undefined

**Comment by pablo2909** [comment]
Apologies I forgot to copy f:

```python
@jax.jit
def f(x, y):
    h = jnp.linalg.norm(x) * jnp.linalg.norm(y)
    for _ in range(20):
        x *= jnp.sin(h*x)
    return x
```

**Comment by patrick-kidger** [comment]
I think what's going on here is that for one of your cases, the derivative could be determined to be *symbolically* zero. See the documentation for [eqx.filter_jvp](https://docs.kidger.site/equinox/api/filtering/transformations/#equinox.filter_jvp), in which it mentions that it will return `None` if it was able to determine that a gradient is a symbolic zero -- that is to say, unconditionally zero regardless of what inputs you provide.

(`filter_vjp` isn't quite as smart -- it'd probably be doable to implement that there but it would be a bit more work)

**Comment by pablo2909** [comment]
That makes sense , thank you ! A bit more info on my use case; the object I get back from `jvpvjp` is added to an other pytree of the same structure (I'm playing a bit with the gradients) but adding a `jax.Array` to a `None` value throws an error. How should we then handle it ?

Thanks a lot for the help

**Comment by patrick-kidger** [comment]
No worries!

As for adding them together, you'd need to add a case for `None` inside your `tree_map`. If the `None`-containing PyTree is also the one used as the reference for the iteration (the first pytree argument to `tree_map`) then you'll also need to mark that `None` should be treated as a leaf:
```python
# must have the same structure when treating Nones as leaves
tree1 = [1.0, 2.0]
tree2 = [3.0, None]

def add1(x, y):
    if y is None:
        return x
    else:
        return x + y

result = jax.tree_util.tree_map(add1, tree1, tree2)

### or

def add2(x, y):
    if x is None:
        return y
    else:
        return x + y

result = jax.tree_util.tree_map(add1, tree2, tree1, is_leaf=lambda x: x is None)

### or

result = eqx.apply_updates(tree1, tree2)
```
And having now explained what it does under the hood -- this is such a common scenario (typically when adding gradient updates to a model) that Equinox ships with the `apply_updates` helper function for exactly this purpose.

**Comment by pablo2909** [comment]
Thank you !
